Configuration/src/main/java/io/deephaven/configuration/Configuration.java

```diff
@@ -4,61 +4,163 @@
+     */
+    private static class DefaultConfiguration extends Configuration {
+        DefaultConfiguration(final @NotNull Supplier<ConfigurationContext> contextSupplier) {
+            super(contextSupplier);
+        }
+
+        @SuppressWarnings("unused")
+        public String getPropertyNullable(String propertyName) {
+            return properties.getProperty(propertyName);
+        }
+    }
+
+    /**
+     * A Named configuration that loads values from the file defined by the property `Configuration.name.rootFile`
+     */
+    private static class NamedConfiguration extends Configuration {
+        private final String name;
+
+        NamedConfiguration(final @NotNull String name, final @NotNull Supplier<ConfigurationContext> contextSupplier) {
+            super(contextSupplier);
+            this.name = name;
+        }
+
+        @Override
+        String determinePropertyFile() {
+            final String propFile = System.getProperty("Configuration." + name + ".rootFile");
+            if (propFile == null) {
+                throw new ConfigurationException("No property file defined for named configuration " + name);
+            }
+
+            return propFile;
+        }
+    }
+
     /**
      * Get the default Configuration instance.
      *
      * @return the single instance of Configuration allowed in an application
      */
     public static Configuration getInstance() {
-        if (INSTANCE == null) {
-            INSTANCE = new NullableConfiguration();
+        if (DEFAULT == null) {
+            DEFAULT = new DefaultConfiguration(DefaultConfigurationContext::new);
+            init(DEFAULT);
         }
-        return INSTANCE;
+        return DEFAULT;
     }
 
-    @SuppressWarnings("UnusedReturnValue")
-    public static Configuration newConfigurationForTesting() {
-        return new Configuration();
+    /**
+     * Get the {@link Configuration} for the specified name.
+     *
+     * @param name the name of the configuration to load
+     * @return the named configuration.
+     * @throws ConfigurationException if the named configuration could not be loaded.
+     */
+    public static Configuration getNamed(@NotNull final String name) throws ConfigurationException {
+        if (DEFAULT_CONF_NAME.equals(name)) {
+            return getInstance();
+        }
+
+        return NAMED_CONFIGURATIONS.computeIfAbsent(name, (k) -> {
+            try {
+                final Configuration instance = new NamedConfiguration(name, DefaultConfigurationContext::new);
+                init(instance);
+                return instance;
+            } catch (ConfigurationException ex) {
+                throw new ConfigurationException("Unable to load named configuration " + name, ex);
+            }
+        });
     }
 
-    protected Configuration() {
+    private static void init(@NotNull final Configuration configuration) {
```

Comment:
Why don't we want to just do this in the super-class constructor?

Reply:
It can't happen from the super because in the Named variant, name wouldn't have been initialized. but I can just call it from both constructors, so I will do that.